### ISSUE TITLE
Anchors with images not handled correctly

### ISSUE THREAD

**Issue by tim-oe** [wkhtmltopdf/wkhtmltopdf#2276]
if you have an anchor tag that has an image and no text the image is not always clickable:

example http://tomcat.apache.org/
the feather in the top right  has a small area that is clickable on the bottom.
i have internal app that that none of the image links are clickable.

running windows 7 64b
C:\Users\Tim>wkhtmltopdf -V
wkhtmltopdf 0.12.2.1 (with patched qt)

**Comment by stuarteske** [comment]
Same here,

Ubuntu 64.
wkhtmltopdf 0.12.2.1 (with patched qt)

Any suggestions?

**Comment by tim-oe** [comment]
haven't heard a peep for a fix.

**Comment by stuarteske** [comment]
Not sure if this work around will help you, but it fixed my issue.  
```css
a {
    display: block;
}
```
```html
<div class="col-xs-2">
    <a href="http://example.com">
        <img src="https://placeholdit.imgix.net/~text?txtsize=22&bg=e8117f&txtclr=ffffff&txt=Social&w=100&h=100" class="img-responsive img-rounded">
    </a>
</div>
````

**Comment by tim-oe** [comment]
so you had to change your html to get it working?

**Comment by stuarteske** [comment]
No, I had to add the style "a {display:block}" to the pdf stylesheet. I was lucky, I didn't have to change the markup.

**Comment by tim-oe** [comment]
ok thanks, i'll give it a try.

**Comment by el-timm** [comment]
I am still having this issue on 0.12.4.  Using the following markup, only the bottom portion of the images are linked in the PDF.

```html
<a href="https://farm4.staticflickr.com/3748/9783751846_11ac9076a7.jpg"><img src="https://farm4.staticflickr.com/3748/9783751846_11ac9076a7_t.jpg" alt=""></a>
<a href="https://farm4.staticflickr.com/3727/9783750986_47c199bc46.jpg"><img src="https://farm4.staticflickr.com/3727/9783750986_47c199bc46_t.jpg" alt=""></a>
<a href="https://farm4.staticflickr.com/3731/9783742834_a1156a9f8a.jpg"><img src="https://farm4.staticflickr.com/3731/9783742834_a1156a9f8a_t.jpg" alt=""></a>
<a href="https://farm3.staticflickr.com/2816/9640881938_91177c7cb5.jpg"><img src="https://farm3.staticflickr.com/2816/9640881938_91177c7cb5_t.jpg" alt=""></a>
```
Also note, this appears to be a duplicate of #2404

**Comment by PhilterPaper** [comment]
The entire Apache logo and title (in the web page) are clickable in my FF browser (205 wide x 83 high), although I have heard reports on some products (such as osCommerce) that only a small portion of some images are clickable (in a browser). That would suggest that there may be some problem in WebKit that it's not making the whole image clickable.

**Comment by PhilterPaper** [comment]
It's hard to tell. If a browser uses the same version of WebKit as wkHTMLtoPDF does (which I understand to be a fairly old one), it might show the same issue, if it's a WebKit problem. So, you might find some browsers that don't allow clicking over the entire logo image, in the same manner as wkHTMLtoPDF, if they share the same WebKit _and_ that's the party with the problem.

**Comment by tekoyaki** [comment]
Also having same issue on 0.12.4.

wkhtmltopdf https://wkhtmltopdf.org/ test.pdf
wkhtmltopdf https://www.google.com test2.pdf (this one's more obvious)

Generated PDF: https://www.pdf-archive.com/2017/05/09/google2/google2.pdf

Only the bottom of the image is clickable. Using OSX 10.11.6.

**Comment by hh2k** [comment]
I can confirm display: block works for us too.

**Comment by NathanCH** [comment]
Like others have said, display block "solves" the link issue but has a lot of side effects. You're turning an inline element into a block element. Therefore you could not have text and images.

Display inline-block works better but it's still problematic. WK will try to insert a line-break occasionally. I'm still debugging the issue but this is the best solution I've come up with so far:

    body  a {
      display: inline-block;
      white-space: nowrap;
    }

Display property will make the images clickable, and the white-space property prevents unnecessary line-breaks. But like with anything in PDFs, there are always exceptions.

**Comment by peca11** [comment]
Same problem here!  Using version 0.12.4. on Ubuntu 16.04.3 

I agree with @PhilterPaper.

It seems that the WebKit QT is using does not process display: "inline" for anchor tags the same way other browsers do.

The only way I got it to work was to use "inline-block" in combination with some js that changed anchors css (to mimic others browsers behaviour) when exporting a pdf.

**Comment by ohboyd** [comment]
Only part of my image was clickable, using the wicked_pdf_image_tag, so this was the solution that worked for me. Just wanted to add it since it took me forever to find this solution on here, and maybe other people will be able use this in the future.

Also, I'm using haml, which is just shorthand for those normal tags you'd use–`%div.image-box` is the same as `<div class='image-box'>other stuff</div>`)

```haml
%div.image-box
  %a{href: "#{client.payment_url}" }
    %img= wicked_pdf_image_tag 'paypal_pay_now.png'
```
```css
a {
  display: block;
}

.image-box {
  margin: 10px 0px;
  float: right;
  padding-right: 35px;
}
```

**Comment by lacripta** [comment]
> Like others have said, display block "solves" the link issue but has a lot of side effects. You're turning an inline element into a block element. Therefore you could not have text and images on the same line.
> 
> Display inline-block works better but it's still problematic. WK will try to insert a line-break occasionally. I'm still debugging the issue but this is the best solution I've come up with so far:
> 
> ```
> a {
>   display: inline-block;
>   white-space: nowrap;
> }
> ```
> 
> Display property will make the images clickable, and the white-space property prevents unnecessary line-breaks caused by inline-block. But like with anything in PDFs, there are always problems. This solution prevents multi-line linked text.

i was having the same issue and this style almost solved it, my images where positioned with an absolute coordinate, after trying a lot i found that moving the style from <img> to <a> solved the issue. just set the anchor at the desired position, the image will move with it, and the whole image area is now clickable

**Comment by tim-oe** [comment]
lacripta 
can you clarify "i found that moving the style from to solved the issue"

**Comment by el-timm** [comment]
@tim-oe looks like @lacripta pasted in HTML tags, which aren't getting escaped in places like the email notification.

e.g. @lacripta 's comment reads:
> i found that moving the style from  to solved the issue

but what was actually posted is:
> i found that moving the style from &lt;img&gt; to &lt;a&gt; solved the issue

Ultimately it looks like they are posting another version of the inline-block workaround.

**Comment by lacripta** [comment]
@tim-oe sorry i didnt reply, just edited my comment, it was an issue with tags not scaped.
@el-timm indeed is another workaround of inline-block, the only difference is that my image had an absolute position, if you try to render the pdf using an image positioned in any place it won't be clickable. the anchor is positioned at the start of the page, but the image will be at the desire position. i though that this could help some else, it was a simple situation of position the anchor tag surrounding the image and not the image itself, in my code example its shown the before and after.

**Comment by tim-oe** [comment]
ok, thanks for the update
